tide-cors/src/middleware.rs

```diff
@@ -0,0 +1,353 @@
+//! Cors middleware
+
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use http::header::HeaderValue;
+use tide_core::{
+    middleware::{Middleware, Next},
+    Context, Response,
+};
+
+/// Middleware for CORS
+///
+/// # Example
+///
+/// ```rust
+///use http::header::HeaderValue;
+///use tide::middleware::CorsMiddleware;
+///
+///CorsMiddleware::new()
+///    .allow_origin(HeaderValue::from_static("*"))
+///    .allow_methods(HeaderValue::from_static("GET, POST, OPTION"))
+///    .allow_credentials(false)
+///    .echo_back_origin(true);
+/// ```
+#[derive(Clone, Debug, Hash)]
+pub struct CorsMiddleware {
+    allow_credentials: HeaderValue,
+    allow_headers: HeaderValue,
+    allow_methods: HeaderValue,
+    allow_origin: HeaderValue,
+    expose_headers: HeaderValue,
+    echo_back_origin: bool,
+    max_age: HeaderValue,
+}
+
+pub const DEFAULT_MAX_AGE: &str = "86400";
+pub const DEFAULT_METHODS: &str = "GET, POST, OPTIONS";
+pub const EMPTY_HEADER: &str = "";
+pub const WILDCARD: &str = "*";
+
+impl CorsMiddleware {
+    /// Creates a new Cors middleware.
+    pub fn new() -> Self {
+        Self {
+            allow_credentials: HeaderValue::from_static("false"),
+            allow_headers: HeaderValue::from_static(WILDCARD),
+            allow_methods: HeaderValue::from_static(DEFAULT_METHODS),
+            allow_origin: HeaderValue::from_static(WILDCARD),
+            expose_headers: HeaderValue::from_static(EMPTY_HEADER),
+            echo_back_origin: false,
+            max_age: HeaderValue::from_static(DEFAULT_MAX_AGE),
+        }
+    }
+
+    /// Set allow_credentials and return new CorsMiddleware
+    pub fn allow_credentials(mut self, allow_credentials: bool) -> Self {
+        self.allow_credentials = HeaderValue::from_str(&allow_credentials.to_string()).unwrap();
+        self
+    }
+
+    /// Set allow_headers and return new CorsMiddleware
+    pub fn allow_headers(mut self, headers: impl Into<HeaderValue>) -> Self {
```

Comment:
Also, just noticed - curious why you switched out the generic implementations in surf with `impl` here?

Reply:
@prasannavl  I think that using 'impl' is more readable than generic notation.
Also, the type is clear when you look at the arguments (in this case 'headers'). When generic, you need to know what 'S' is.

```rust
 pub fn allow_headers<S: Into<HeaderValue>>(mut self, headers: S) -> Self {

pub fn allow_headers(mut self, headers: impl Into<HeaderValue>) -> Self {
```

I felt that 'impl' is often used in tide.
https://github.com/rustasync/tide/blob/master/tide-core/src/app.rs#L223

Reply:
aha! cool. Thanks for the note! -- I'll create an issue for this soon so we can gather consensus. But marking this as resolved, since it was just a question of curiosity :)

Reply:
@k-nasa in general it's preferable for library code to use generic arguments because `impl trait` doesn't allows turbofish application of arguments yet.

I agree `impl` is more readable, but think we should generally lean to using generic parameters for Tide until the restraints around turbofish + impl trait are resolved.

Reply:
OK. I did it